internal/service/athena/data_catalog.go

```diff
@@ -0,0 +1,254 @@
+	conn := meta.(*conns.AWSClient).AthenaConn
+
+	if d.HasChangesExcept("tags", "tags_all") {
+		input := &athena.UpdateDataCatalogInput{
+			Name:        aws.String(d.Id()),
+			Type:        aws.String(d.Get("type").(string)),
+			Description: aws.String(d.Get("description").(string)),
+		}
+
+		if d.HasChange("parameters") {
+			parameters := map[string]*string{}
+			if v, ok := d.GetOk("parameters"); ok {
+				if m, ok := v.(map[string]interface{}); ok {
+					parameters = flex.ExpandStringMap(m)
+				}
+			}
+			input.Parameters = parameters
+		}
+
+		log.Printf("[DEBUG] Updating Athena Data Catalog (%s)", d.Id())
+
+		if err := input.Validate(); err != nil {
+			return diag.Errorf("Error validating Athena Data Catalog (%s): %s", d.Id(), err)
+		}
+
+		_, err := conn.UpdateDataCatalogWithContext(ctx, input)
+
+		if err != nil {
+			return diag.Errorf("error updating Athena Data Catalog (%s): %s", d.Id(), err)
+		}
+	}
+
+	if d.HasChange("tags_all") {
+		o, n := d.GetChange("tags_all")
+
+		log.Printf("[DEBUG] Updating Athena Data Catalog (%s) tags", d.Id())
+		if err := UpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return diag.Errorf("error updating Athena Data Catalog (%s) tags: %s", d.Id(), err)
+		}
+	}
+
+	return resourceDataCatalogRead(ctx, d, meta)
+}
+
+func resourceDataCatalogDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).AthenaConn
+
+	log.Printf("[DEBUG] Deleting Athena Data Catalog: (%s)", d.Id())
+
+	_, err := conn.DeleteDataCatalogWithContext(ctx, &athena.DeleteDataCatalogInput{
+		Name: aws.String(d.Id()),
+	})
+
+	if tfawserr.ErrCodeEquals(err, athena.ErrCodeResourceNotFoundException) {
+		return nil
+	}
+
+	if err != nil {
+		return diag.Errorf("error deleting Athena Data Catalog (%s): %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func resourceDataCatalogRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).AthenaConn
+	defaultTagsConfig := meta.(*conns.AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*conns.AWSClient).IgnoreTagsConfig
+
+	input := &athena.GetDataCatalogInput{
+		Name: aws.String(d.Id()),
+	}
+
+	dataCatalog, err := conn.GetDataCatalogWithContext(ctx, input)
+
+	// If the resource doesn't exist, the API returns a `ErrCodeInvalidRequestException` error.
+	if !d.IsNewResource() && tfawserr.ErrMessageContains(err, athena.ErrCodeInvalidRequestException, "was not found") {
+		log.Printf("[WARN] Athena Data Catalog (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return diag.Errorf("error reading Athena Data Catalog (%s): %s", d.Id(), err)
+	}
+
+	d.Set("description", dataCatalog.DataCatalog.Description)
+	d.Set("type", dataCatalog.DataCatalog.Type)
+
+	// NOTE: This is a workaround for the fact that the API sets default values for parameters that are not set.
```

Comment:
The API allows us to set the `parameters` as following:

```terraform
 parameters = {
    "function"   = "arn:aws:lambda:eu-central-1:837424938642:function:not-important-lambda-function"
  }
  ```
 
But behind the scenes the parameters are changed and set to `record-function, metadata-function and catalog` (depending on which data catalog type you pick).

At the moment I work around it by omitting those from the state and only setting in the state what the user supplies. Otherwise I keep running into the below issue (where it's trying to delete the parameters that are created behind the scenes)..

Any thoughts?

```
Terraform will perform the following actions:

  # aws_athena_data_catalog.example will be updated in-place
  ~ resource "aws_athena_data_catalog" "example" {
        id          = "tf-test-catalog"
        name        = "tf-test-catalog"
      ~ parameters  = {
          - "catalog"           = "tf-test-catalog" -> null
          - "metadata-function" = "arn:aws:lambda:eu-central-1:837424938642:function:not-important-lambda-function" -> null
          - "record-function"   = "arn:aws:lambda:eu-central-1:837424938642:function:not-important-lambda-function" -> null
            # (1 unchanged element hidden)
        }
        tags        = {}
        # (4 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Reply:
Hmm, do you perhaps have some advice @ewbankkit ?

Reply:
I tweaked this to read back the values of parameter keys that the practitioner has configured.